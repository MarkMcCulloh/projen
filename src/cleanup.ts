import * as path from 'path';
import * as fs from 'fs-extra';
import { existsSync, readFileSync } from 'fs-extra';
import * as glob from 'glob';
import { PROJEN_DIR, PROJEN_MARKER } from './common';
import * as logging from './logging';

export const FILE_MANIFEST = `${PROJEN_DIR}/files.json`;

export function cleanup(dir: string, exclude: string[]) {
  const fileManifestPath = path.resolve(dir, FILE_MANIFEST);
  try {
    if (existsSync(fileManifestPath)) {
        const fileFile = JSON.parse(readFileSync(fileManifestPath, 'utf-8'));
        for (const file of fileFile.files) {
          fs.removeSync(path.resolve(dir, file));
        }
      }
  } catch (e) {
    logging.warn(`warning: failed to clean up generated files using file manifest: ${e.stack}`);
  }

  // manifest file is invalid or not present
  try {
    for (const f of findGeneratedFiles(dir, exclude)) {
      fs.removeSync(f);
    }
  } catch (e) {
    logging.warn(`warning: failed to clean up generated files: ${e.stack}`);
  }
}

function findGeneratedFiles(dir: string, exclude: string[]) {
  const ignore = [...readGitIgnore(dir), 'node_modules/**', ...exclude, '.git/**'];

  const files = glob.sync('**', {
    ignore,
    cwd: dir,
    dot: true,
    nodir: true,
    absolute: true,
  });

  const generated = new Array<string>();

  for (const file of files) {

    const contents = fs.readFileSync(file, 'utf-8');

    if (contents.includes(PROJEN_MARKER)) {
      generated.push(file);
    }
  }

  return generated;
}

function readGitIgnore(dir: string) {
  const filepath = path.join(dir, '.gitignore');
  if (!fs.pathExistsSync(filepath)) {
    return [];
  }

  return fs.readFileSync(filepath, 'utf-8')
    .split('\n')
    .filter(x => x?.trim() !== '')
    .filter(x => !x.startsWith('#') && !x.startsWith('!'))
    .map(x => x.replace(/^\//, '')) // remove "/" prefix
    .map(x => `${x}\n${x}/**`)
    .join('\n')
    .split('\n');
}

