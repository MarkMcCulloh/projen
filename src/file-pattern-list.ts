export interface FilePatternListOptions {
  readonly includedPatterns?: string[];
  readonly excludedPatterns?: string[];
  readonly preferExcluded?: boolean;
}

export class FilePatternList {
  private readonly _includedPatterns = new Array<string>();
  private readonly _excludedPatterns = new Array<string>();
  private readonly _preferExcluded;

  public get patterns() {
    return this._includedPatterns.concat(this._excludedPatterns);
  }
  public get excludedPatterns() {
    return this._excludedPatterns;
  }
  public get includedPatterns() {
    return this._includedPatterns;
  }

  constructor(options?: FilePatternListOptions) {
    this._preferExcluded = options?.preferExcluded ?? false
    options?.excludedPatterns?.forEach(this.addPattern)
    options?.includedPatterns?.forEach(this.addPattern)
  }

  /**
   * Add file patterns. Files that match this pattern will be included. If the
   * pattern starts with a negation mark `!`, files that match will _not_ be
   * included.
   *
   * Comment lines (start with `#`) are ignored.
   *
   * @param patterns Ignore patterns.
   */
  public addPatterns(...patterns: string[]) {
    for (const pattern of patterns) {
      this.addPattern(pattern);
    }
  }

  /**
   * Removes patterns previously added.
   *
   * If `addPatterns()` is called after this, the pattern will be added again.
   *
   * @param patterns patters to remove.
   */
  public removePatterns(...patterns: string[]) {
    for (const p of patterns) {
      remove(this._includedPatterns, p);
      remove(this._excludedPatterns, p);
    }
  }

  /**
   * Ignore the files that match these patterns.
   * @param patterns The patterns to match.
   */
  public exclude(...patterns: string[]) {
    for (let pattern of patterns) {
      if (!pattern.startsWith('!')) {
        pattern = '!' + pattern;
      }

      this.addPatterns(pattern);
    }
  }

  private addPattern(pattern: string) {
    // skip comments
    if (pattern.startsWith('#')) {
      return;
    }

    const isExclude = pattern.startsWith('!')
    const isDirectory = pattern.endsWith('/')
    const opposite = isExclude ? pattern.slice(1) : `!${pattern}`;
    if(isExclude) {
      remove(this._excludedPatterns, pattern);
      if(this._preferExcluded) {
        remove(this._includedPatterns, opposite);
      }
    } else {
      remove(this._includedPatterns, pattern);
      if(!this._preferExcluded) {
        remove(this._excludedPatterns, opposite);
      }
    }

    if (isDirectory) {
      const prefix = opposite;
      for (const p of [...this._includedPatterns]) {
        if (p.startsWith(prefix)) {
          remove(this._includedPatterns, p);
        }
      }
      for (const p of [...this._excludedPatterns]) {
        if (p.startsWith(prefix)) {
          remove(this._excludedPatterns, p);
        }
      }
    }
  }
}

// O(n) hooray!
function remove<T>(arr: T[], value: any) {
  const idx = arr.indexOf(value);
  if (idx >= 0) {
    arr.splice(idx, 1);
  }
}
