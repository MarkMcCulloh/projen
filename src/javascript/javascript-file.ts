import { js } from "js-beautify";
import { Project } from "../project";
import { TextFile, TextFileOptions } from "../textfile";

function renderStatement(jsStatement: IJavascriptStatement) {
  if (jsStatement.raw !== undefined) {
    return jsStatement.raw;
  }

  // Validate arguments
  if (!jsStatement.code) {
    throw new Error("`code` must be provided if `raw` is not set");
  }
  if (jsStatement.callWith) {
    if (!jsStatement.asFunction) {
      throw new Error("`asFunction` must be true if `callWith` is set");
    }
  }
  if (jsStatement.exportDefault && jsStatement.exportName !== undefined) {
    throw new Error("Cannot specify `exportName` when `exportDefault` is true");
  }

  let fString = jsStatement.code.toString().trim();

  if (!jsStatement.asFunction) {
    fString = fString.substring(
      fString.indexOf("{") + 1,
      fString.lastIndexOf("}")
    );
  } else {
    if (jsStatement.callWith) {
      fString = `(${fString})(${
        jsStatement.callWith.map((c) => `\`${c.toString()}\``).join(", ") ?? ""
      })`;
    }
  }
  if (jsStatement.setToValue) {
    fString = `const ${jsStatement.setToValue} = ${fString};`;
  }
  if (jsStatement.exportDefault) {
    fString = `module.exports = ${fString};`;
  } else if (jsStatement.exportName) {
    fString = `exports.${jsStatement.exportName} = ${fString};`;
  }
  if (jsStatement.prefix) {
    fString = jsStatement.prefix + fString;
  }
  if (jsStatement.postfix) {
    fString = fString + jsStatement.postfix;
  }
  return fString;
}

export interface IJavascriptStatement {
  /**
   * Creates a const with the given name and sets it equal to the given function
   * asFunction must also be true.
   */
  setToValue?: string;

  /**
   * Use the given function declaration when written to a file
   *
   * @default false Function declaration is stripped
   */
  asFunction?: boolean;

  /**
   * Text added before the given function text
   */
  prefix?: string;

  /**
   * Text added after the given function text
   */
  postfix?: string;

  /**
   * exports.{exportName} = {value}
   */
  exportName?: string;

  /**
   * module.exports = {value}
   */
  exportDefault?: boolean;

  /**
   * Overrides all other options
   */
  raw?: string;

  /**
   * Represents the function as a call with the given string values. Values will be .toString()'d.
   * asFunction must also be true.
   *
   * To call the function with no values, use an empty array.
   */
  callWith?: any[];

  /**
   * Function containing relevant code.
   * Arguments can be used to simulate the variables that would be available
   *
   * Note: This Javascript is not executed in Projen, and it should only use values that would be available in the scope it is executed in.
   */
  code?(..._: any): any;
}

export interface IIFEStatement {
  callWith?: any[];
  fn(..._: any): any;
}

export interface IJavascriptFileOptions {
  /**
   * options passed to underlying TextFile
   */
  textFileOptions?: TextFileOptions;
}

/**
 * A Javascript file
 */
export class JavascriptFile extends TextFile {
  private statements: IJavascriptStatement[];

  constructor(
    project: Project,
    path: string,
    options?: IJavascriptFileOptions
  ) {
    super(project, path, options?.textFileOptions);

    this.statements = [];
  }

  addNewline() {
    this.addStatement({
      raw: "",
    });
  }

  addIIFE(iifeStatement: IIFEStatement): IJavascriptStatement {
    const jsStatement: IJavascriptStatement = {
      ...iifeStatement,
    };

    this.addStatement(jsStatement);

    return jsStatement;
  }

  /**
   * Add new Javascript statement to the end of the list
   */
  addStatement(jsStatement: IJavascriptStatement): void {
    this.statements.push(jsStatement);
  }

  /**
   * Add new Javascript statement to the beginning of the list
   */
  prependStatement(jsStatement: IJavascriptStatement): void {
    this.statements.unshift(jsStatement);
  }

  preSynthesize() {
    if (this.readonly) {
      this.addLine(`// ${this.marker}`);
    }

    const code = this.statements.map((f) => renderStatement(f)).join("\n");

    this.addLine(
      js(code, {
        indent_level: 0,
        indent_empty_lines: false,
        indent_size: 2,
        end_with_newline: true,
        preserve_newlines: true,
        max_preserve_newlines: 2,
      })
    );
  }
}
