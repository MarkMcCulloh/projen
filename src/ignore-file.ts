import { FileBase, IResolver } from './file';
import { FilePatternList } from './file-pattern-list';
import { Project } from './project';


export class IgnoreFile extends FileBase {
  private readonly _filePatternList = new FilePatternList();

  constructor(project: Project, filePath: string) {
    super(project, filePath, { editGitignore: filePath !== '.gitignore' });
  }

  /**
   * Add ignore patterns. Files that match this pattern will be ignored. If the
   * pattern starts with a negation mark `!`, files that match will _not_ be
   * ignored.
   *
   * Comment lines (start with `#`) are ignored.
   *
   * @param patterns Ignore patterns.
   */
  public addPatterns(...patterns: string[]) {
    this._filePatternList.addPatterns(...patterns);
  }

  /**
   * Removes patterns previously added from the ignore file.
   *
   * If `addPattern()` is called after this, the pattern will be added again.
   *
   * @param patterns patterns to remove.
   */
  public removePatterns(...patterns: string[]) {
    this._filePatternList.removePatterns(...patterns);
  }
  /**
   * Ignore the files that match these patterns.
   * @param patterns The patterns to match.
   */
  public exclude(...patterns: string[]) {
    return this._filePatternList.addPatterns(...patterns);
  }

  /**
   * Always include the specified file patterns.
   * @param patterns Patterns to include in git commits.
   */
  public include(...patterns: string[]) {
    return this._filePatternList.exclude(...patterns);
  }

  protected synthesizeContent(resolver: IResolver): string | undefined {
    const lines = [
      `# ${FileBase.PROJEN_MARKER}`,
      ...this._filePatternList.patterns,
    ];

    return `${resolver.resolve(lines).join('\n')}\n`;
  }
}