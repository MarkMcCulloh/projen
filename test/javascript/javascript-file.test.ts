import { JavascriptFile } from "../../src/javascript";
import { synthSnapshot, TestProject } from "../util";

test("can add javascript file component", () => {
  const prj = new TestProject();

  const file = new JavascriptFile(prj, "cool.config.js");

  file.addStatement({
    code: () => {
      console.log("Hello World");
    },
  });

  const out = synthSnapshot(prj);
  const outFile = out["cool.config.js"];
  console.log(outFile);

  expect(outFile).toBe(`// ${file.marker}\nconsole.log('Hello World');\n`);
  expect(outFile).toMatchSnapshot();
});

test("can add javascript file component with various options", () => {
  const prj = new TestProject();

  const file = new JavascriptFile(prj, "cool.config.js");

  file.addStatement({
    code: () => {
      console.log("Hello World");
    },
  });

  file.addNewline();

  file.addStatement({
    code: (name) => {
      return `Hello ${name}`;
    },
    asFunction: true,
    callWith: [prj.name],
    setToValue: "greeting",
  });

  file.addStatement({
    code: (greeting) => {
      console.log(greeting);
    },
    prefix: "// do a log\n",
    postfix: "\n// did a log",
  });

  const out = synthSnapshot(prj);
  const outFile = out["cool.config.js"];

  console.log(outFile);

  expect(outFile).toContain("// do a log");
  expect(outFile).toContain("// did a log");
  expect(outFile).toMatchSnapshot();
});
