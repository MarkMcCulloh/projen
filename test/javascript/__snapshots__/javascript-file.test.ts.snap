// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add javascript file component 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
console.log('Hello World');
"
`;

exports[`can add javascript file component with various options 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
console.log('Hello World');

const greeting = ((name) => {
  return \`Hello \${name}\`;
})(\`my-project\`);
// do a log

console.log(greeting);

// did a log
"
`;
